Documentation template and unit test templates

magrathj opened the pull request:
Added template documentation to  shinyforms (still not complete but allows for documentation to be update using the defined format), basic unit tests on three exported functions (similarly, will keep updating once functions are built out more) and lastly added code for createFormApp.

magrathj committed:
type: docs

Added roxygen template documentation to the code.  Will fill out documentation in more detail after editing the code.

magrathj committed:
type: feature

Added unit tests for createFormApp and added simple shiny code to run createFormApp like shiny::runExample() function

resolves: issue #9

magrathj committed:
type: test

Added basic unit tests for functions being exported and added extra documention.

daattali commented on DESCRIPTION:
```diff
@@ -12,12 +12,14 @@ Imports:
-Remotes:  rstudio/shiny
+    rmarkdown,
+    testthat
+Remotes: rstudio/shiny
```
shiny can move into Imports instead of Remotes :)

daattali commented on R/shinyform.R:
```diff
@@ -11,13 +21,31 @@ STORAGE_TYPES <- list(
+
+#' @title labelMandatory
+#' 
+#' @description Adds a mandaotory star to a labelled question.
```
typo

daattali commented on DESCRIPTION:
```diff
@@ -12,12 +12,14 @@ Imports:
     shiny (>= 0.13.2.9003),
-    shinyjs (>= 0.5.2)
+    shinyjs (>= 0.5.2),
+    checkmate
```
I prefer not to introduce this package, unless it has some uses that really can't be easily done without it

daattali reviewed (commented):
Thanks a lot for the initiative!

A few comments:

- For all the examples in the documentation, can you wrap the entire example code in `if (interactive()) { ... }` to ensure the examples only run in interactive environments

- There's no need to document all the internal functions

- For all the other functions that do need documentation, do you mind changing the format to more closely match http://r-pkgs.had.co.nz/man.html ? There's no need to explicitly specify `@title` and `@description` and `@details` sections. The title should not be the name of the function, it should be a very short few-word description

magrathj committed:
type: fix

Thanks for the feedback @daattali.

Implement all comments and conversations in review feedback.

Changed the documentation for internal functions to just basic comments to help me remember what each function does - tell me if you still want me to remove it as well?

Removed checkmate package and added own testList function

Added interactive to code examples and removed @title, @details and @descriptions in documentation.

magrathj committed:
type:fix

realised i forgot to export the storage types

daattali commented on DESCRIPTION:
```diff
@@ -12,12 +12,13 @@ Imports:
     shiny (>= 0.13.2.9003),
-    shinyjs (>= 0.5.2)
+    shinyjs (>= 0.5.2),
+    rstudio/shiny
```
This line should be removed

magrathj replied:
removed this line - will be in the next commit

daattali commented on NAMESPACE:
```diff
@@ -1,6 +1,5 @@
 # Generated by roxygen2: do not edit by hand
 
-export(STORAGE_TYPES)
```
For some reason STORAGE_TYPES was removed from the NAMESPACE. Roxygen likely just needs to run again

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
 library(shiny)
 
+# test whethere list is empty or not a list
```
Typo "whethere". Also the wording is grammatically incorrect and a bit awkward. "Test whether a given object is a valid non-empty list"

magrathj replied:
Cheers, just updated - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
 library(shiny)
 
+# test whethere list is empty or not a list
+# @para listname a potential list to verify
```
typo "para"

magrathj replied:
updated - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
 
+# test whethere list is empty or not a list
+# @para listname a potential list to verify
+# returns FALSE if not a list or empty
```
If the previous line is following the roxygen tags with the `@` sign, this line might as well also do that. Also, the phrase is worded as an incomplete sentence. "TRUE if the given object is a non-empty list, FALSE otherwise"

magrathj replied:
updated sentence to read "TRUE if the given object is a non-empty list, FALSE otherwise" and changed para to param - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
+# @para listname a potential list to verify
+# returns FALSE if not a list or empty
+testList <- function(listname){
+  return(!is.null(listname) & 
```
Is `listname` meant to be a single object, or a vector? Assuming it's meant to be a single object, you need to use `&&` rather than `&`

magrathj replied:
Yeah thanks, meant to be a single object. Have updated - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
+testList <- function(listname){
+  return(!is.null(listname) & 
+         length(listname) != 0 &
+         class(listname) == "list")
```
I think it's safer to check if `"list" %in% class(listname)` because the object may have multiple classes

magrathj replied:
awesome thanks, just updated - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
+
+
+# A list of all the available storage types for shinyforms.
+# List: FLATFILE, SQLITE, MYSQL, MONGO, GOOGLE_SHEETS, DROPBOX, AMAZON_S3
```
No need for this line that specifies all the types - since that's the exact purpose of this variable. Trying to list out the storage types in a comment means that any future changes to the variable have to be remembered to be modified in the comment as well, which is an unnecessary place to introduce a logical conflict

magrathj replied:
removed line - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -1,5 +1,18 @@
+
+# A list of all the available storage types for shinyforms.
+# List: FLATFILE, SQLITE, MYSQL, MONGO, GOOGLE_SHEETS, DROPBOX, AMAZON_S3
+# A list of storage types available to shinyform.
```
This line is unnecessary, it's almost a duplicate of the above

daattali commented on R/shinyform.R:
```diff
@@ -11,13 +24,20 @@ STORAGE_TYPES <- list(
+
+
+# Adds a mandatory star to a labelled question.
+# label A string representing the mandatory question.
```
It isn't clear that the word `label` on the second line is referring to the name of a parameter. It's ok to use `@param` to make that clear, or to remove the line describing the param entirely

magrathj replied:
added param - will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -73,15 +117,65 @@ loadDataFlatfile <- function(storage) {
 #' @export
 formUI <- function(formInfo) {
+  if (!testList(formInfo)) {
+    stop("`formInfo` missing from shinyApp")
```
This error message is incorrect. It should say "formInfo is not a valid list"

magrathj replied:
updated all stop messages to read "formInfo is not a valid list"- will be in the next commit

daattali commented on R/shinyform.R:
```diff
@@ -193,11 +287,56 @@ formUI <- function(formInfo) {
 #' @export
 formServer <- function(formInfo) {
+  if (!testList(formInfo)) {
+    stop("`formInfo` missing from shinyApp")
```
same as above

daattali reviewed (changes requested):
Thanks for addressing it all! I have a few more comments when you have time.

magrathj committed:
type: fix

Updated changes from review. Fixed typos, re-ran devtool:document(), remove specific lines, changes to testList function
